scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -1403,93 +1380,42 @@ private void processChangeCipherSpecRecord(Record record, Connection connection)
 	private void processHandshakeRecord(final Record record, final Connection connection) {
 		LOGGER.debug("Received {} record from peer [{}]", record.getType(), record.getPeerAddress());
 		try {
-			if (connection.hasOngoingHandshake()) {
-				DTLSSession handshakeSession = connection.getOngoingHandshake().getSession();
-				if (handshakeSession.getReadEpoch() == record.getEpoch()) {
-					// evaluate message in context of ongoing handshake
-					record.applySession(handshakeSession);
-				} else if (!record.isNewClientHello()) {
-					// epoch is not the same as the current session so we
-					// can not decrypt the message now. Let handshaker handle it
-					// (it can queue it to deal with it later)
-					connection.getOngoingHandshake().processMessage(record);
-					return;
-				}
-			} else if (connection.hasEstablishedSession()
-					&& connection.getEstablishedSession().getReadEpoch() == record.getEpoch()) {
-				// client wants to re-negotiate established connection's
-				// crypto params evaluate message in context of established session
-				record.applySession(connection.getEstablishedSession());
-			} else if (record.isNewClientHello()) {
-				// client has lost track of existing connection and wants to
-				// negotiate a new connection
-				// in epoch 0 no crypto params have been established yet, thus
-				// we do not need to set a session
-			} else {
-				LOGGER.debug(
-						"Discarding HANDSHAKE message [epoch={}] from peer [{}] which does not match expected epoch(s) [{}]",
-						record.getEpoch(), record.getPeerAddress());
-				return;
+			if (record.isNewClientHello()) {
+				throw new IllegalArgumentException("new CLIENT_HELLO must be processed by processClientHello!");
```

Comment:
is this check more or less the same as the one below (`else` part)?
```java
if (record.getEpoch() > 0) {
	DTLSSession session = connection.getEstablishedSession();
	send(new AlertMessage(AlertLevel.WARNING, AlertDescription.NO_RENEGOTIATION, record.getPeerAddress()),session);
} else {
	LOGGER.error("Unexpected CLIENT_HELLO {}", record.getPeerAddress(),	new IllegalArgumentException("Unexpected CLIENT_HELLO"));
}
```

Reply:
Seems so.

Reply:
does it mean we should remove one of them ?

Reply:
Done.